src/libraries/System.IO.Pipes/src/System/IO/Pipes/PipeValueTaskSource.cs

```diff
@@ -3,14 +3,14 @@
 
 using System.Buffers;
 using System.Diagnostics;
+using System.Runtime.ExceptionServices;
 using System.Runtime.InteropServices;
-using System.Security;
 using System.Threading;
-using System.Threading.Tasks;
+using System.Threading.Tasks.Sources;
 
 namespace System.IO.Pipes
 {
-    internal abstract unsafe class PipeCompletionSource<TResult> : TaskCompletionSource<TResult>
+    internal abstract unsafe class PipeValueTaskSource<TResult> : IValueTaskSource<TResult>, IValueTaskSource
```

Comment:
If memory serves, this implementation and the old completion source used with FileStream were very similar / copy-and-paste.  But we rewrote it almost entirely into the new ValueTaskSource.  Can you please bring this to be much closer in design to https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/IO/Strategies/AsyncWindowsFileStreamStrategy.ValueTaskSource.cs?

Reply:
I have introduced the use of `PreAllocatedOverlapped` and moved the PipeValueTaskSource as a nested class. Is there anything else to bring in?